server/auth/providers/oidc.js

```diff
@@ -0,0 +1,129 @@
+// @flow
+import passport from "@outlinewiki/koa-passport";
+import fetch from "fetch-with-proxy";
+import Router from "koa-router";
+import { Strategy } from "passport-oauth2";
+import accountProvisioner from "../../commands/accountProvisioner";
+import env from "../../env";
+import { OIDCMalformedUserInfoError, AuthenticationError } from "../../errors";
+import passportMiddleware from "../../middlewares/passport";
+import { getAllowedDomains } from "../../utils/authentication";
+import { StateStore } from "../../utils/passport";
+
+const router = new Router();
+const providerName = "oidc";
+const OIDC_DISPLAY_NAME = process.env.OIDC_DISPLAY_NAME || "OpenID Connect";
+const OIDC_CLIENT_ID = process.env.OIDC_CLIENT_ID;
+const OIDC_CLIENT_SECRET = process.env.OIDC_CLIENT_SECRET;
+const OIDC_AUTH_URI = process.env.OIDC_AUTH_URI;
+const OIDC_TOKEN_URI = process.env.OIDC_TOKEN_URI;
+const OIDC_USERINFO_URI = process.env.OIDC_USERINFO_URI;
+const OIDC_TEAM_CLAIM = process.env.OIDC_TEAM_CLAIM || "team_name";
+const OIDC_SCOPES = process.env.OIDC_SCOPES || "";
+const allowedDomains = getAllowedDomains();
+
+export const config = {
+  name: OIDC_DISPLAY_NAME,
+  enabled: !!OIDC_CLIENT_ID,
+};
+
+const scopes = OIDC_SCOPES.split(" ");
+
+Strategy.prototype.userProfile = async function (accessToken, done) {
+  try {
+    const response = await fetch(OIDC_USERINFO_URI, {
+      credentials: "same-origin",
+      headers: {
+        Authorization: `Bearer ${accessToken}`,
+      },
+    });
+
+    try {
+      return done(null, await response.json());
+    } catch (err) {
+      return done(err);
+    }
+  } catch (err) {
+    return done(err);
+  }
+};
+
+if (OIDC_CLIENT_ID) {
+  passport.use(
+    providerName,
+    new Strategy(
+      {
+        authorizationURL: OIDC_AUTH_URI,
+        tokenURL: OIDC_TOKEN_URI,
+        clientID: OIDC_CLIENT_ID,
+        clientSecret: OIDC_CLIENT_SECRET,
+        callbackURL: `${env.URL}/auth/${providerName}.callback`,
+        passReqToCallback: true,
+        scope: OIDC_SCOPES,
+        store: new StateStore(),
+        state: true,
+        pkce: false,
+      },
+
+      // OpenID Connect standard profile claims can be found in the official
+      // specification.
+      // https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+
+      // Non-standard claims may be configured by individual identity providers.
+      // Any claim supplied in response to the userinfo request will be
+      // available on the `profile` parameter
+      async function (req, accessToken, refreshToken, profile, done) {
+        try {
+          const parts = profile.email.split("@");
```

Comment:
I don’t know about using user-provided email to “verify” domain ownership (something like `email_verified` would be more appropriate for it) as it’d be a security risk - you can pass whatever email you’d like at the point of signing up… Probably better to check `aud` or `issuer` of the JWT claim as they’re standard by the OIDC spec!

Reply:
In this case it is more for domain membership (I think). There's another env variable that allows you to limit access to users that have email addresses on a particular domain.

Reply:
If a user completes a OIDC flow w/ your given client, they are already "in the system", meaning they already belong to the domain. If one wishes to *limit* the "pool" of people who can sign into outline, it would be better done by using the *client* to do so - after all, identity providers can configure clients to have access only to certain people/pools/etc, so by authenticating with the locked-down client you can limit who can login to outline.

Reply:
Outline uses many identity providers, some of which are social logins in their nature. This is simply to provide compatibility to an already existing access control mechanism; one doesn't _need_ to specify the `ALLOWED_DOMAINS` variable. That being said, for my own use case I am looking to control user authorization directly from the IdP as you describe, and thus would not be using this mechanism.

Reply:
I think if `issuer` is guaranteed to be a domain it would be okay to use that instead of splitting the email

Reply:
Issuer is a URI